code/go/0chain.net/smartcontract/dbs/event/eventdb.go

```diff
@@ -15,17 +20,81 @@ func NewEventDb(config dbs.DbAccess) (*EventDb, error) {
 		return nil, err
 	}
 	eventDb := &EventDb{
-		Store: db,
+		Store:          db,
+		eBufferChannel: make(chan eventCtx, 100),
+		eChannel:       make(chan eventCtx, 100),
 	}
 
 	if err := eventDb.AutoMigrate(); err != nil {
 		return nil, err
 	}
+	go eventDb.channelBufferIntermediate()
+	go eventDb.addEventWorker()
 	return eventDb, nil
 }
 
+type eventCtx struct {
+	ctx    context.Context
+	events []Event
+}
+
 type EventDb struct {
 	dbs.Store
+	eChannel       chan eventCtx
+	eBufferChannel chan eventCtx
+}
+
+// helps maitaining dynamic buffer for addEventWorker
+func (edb EventDb) channelBufferIntermediate() {
+	buf := make([]eventCtx, 0)
+	for {
+		events, ok := <-edb.eBufferChannel
+		for _, e := range buf {
+			select {
+			case edb.eChannel <- e:
+			default:
+				break
+			}
+			buf = buf[1:]
+		}
+		if ok {
+			select {
+			case edb.eChannel <- events:
+			default:
+				buf = append(buf, events)
+			}
+		}
+	}
+}
+
+// addEventWorker this worker will try to add events unless thery are not added.
+func (edb EventDb) addEventWorker() {
+	for {
+		events := <-edb.eChannel
+		for {
+			newEvents := edb.removeDuplicate(events.ctx, events.events)
+			if err := edb.addEvents(events.ctx, newEvents); err != nil && !strings.Contains(err.Error(), "len(events):0") {
```

Comment:
it might be not a good idea to pass context via event, should we write events to db if this context passed in for example computeState() got cancelled or timeout? I think yes, we better use some new context with timeout here, just to prevent long db connections

Reply:
This is more a product behavior if the case you are mentioning is what we want then why not use context timeout in the new worker itself?
wdyt would that be fine?

Reply:
is there any reason for this context cancellation other than timeout?

Reply:
I did not trace but we do use context.WithCancel
![image](https://user-images.githubusercontent.com/11261548/150656830-9065589c-1827-4eef-9c3a-dd8720da3a68.png)
at some places 
So what if those cancellations are encountered.
Having a disconnected context will make us miss those cancellations

Reply:
unfortunately context is not equal to transaction. we wish to have single transactional scope for block and for event, I mean when block computation is cancelled than to cancel event writing and if block is committed (there is no such a concept yet) then commit transaction in event db. 

I agree that to ensure data consistency we have to implement something similar to distributed transaction, to be honest we don't want to use one since it will be overkill, we might:
- implement some compensation logic for cases when block state computation is cancelled, for example to generate appropriate event and to mark all generated events previously as cancelled
- we can replace event in your solution with some channel (semaphore) to start event writing after it 

the problem with your solution is that we *can* cancel the event writing if context will be cancelled before the writing, but it is not strictly